docker_cli/_utils.py

```diff
@@ -0,0 +1,256 @@
+import io
+import random
+import re
+from shlex import split
+from string import ascii_lowercase, digits
+from subprocess import DEVNULL, CalledProcessError, run
+from threading import Lock
+from typing import Optional, Tuple, Type
+
+from ._image import Image
+from ._shell_cmds import (
+    PackageManager,
+    URLReader,
+    get_package_manager,
+    get_supported_package_managers,
+    get_supported_url_readers,
+    get_url_reader,
+)
+
+
+def universal_tag_prefix() -> str:
+    """
+    Returns a prefix for tags generated by the system.
+
+    Returns
+    -------
+    str
+        The prefix.
+    """
+    return "dcli"
+
+
+def _package_manager_check(image: Image) -> PackageManager:
+    """
+    Returns the package manager present on an image.
+
+    Parameters
+    ----------
+    base : Image
+        The image.
+
+    Returns
+    -------
+    PackageManager
+        The package manager.
+    """
+    package_mgrs = image.check_command_availability(get_supported_package_managers())
+
+    if package_mgrs:
+        package_mgr = get_package_manager(package_mgrs[0])
+    else:
+        ValueError("No recognized package manager found on parent image.")
+
+    return package_mgr
+
+
+def _url_reader_check(
+    image: Image, package_mgr: PackageManager
+) -> Tuple[Type[URLReader], str]:
+    """
+    Return the URL reader on a given image, and a string to install one if there is
+    none.
+
+    Parameters
+    ----------
+    base : Image
+        The image.
+    package_mgr : PackageManager
+        The images package manager.
+
+    Returns
+    -------
+    Tuple[Type[URLReader], str]
+        The installed URL reader and a string to install it if necessary.
```

Comment:
When you have multiple return values, you can document each one individually

```suggestion
    Type[URLReader]
        The installed URL reader.
    str
        A string to install the URL reader if necessary.
```

Reply:
The numpy docstring guide doesn't explicitly require this, but I also recommend giving each return value a name when you're returning more than one thing

```suggestion
    url_reader : Type[URLReader]
        The installed URL reader.
    install_command : str
        A string to install the URL reader if necessary.
```